File: data structure and algorithms/Array/533.lonely-pixel-ii.java
/*
 * @lc app=leetcode id=533 lang=java
 *
 * [533] Lonely Pixel II
 *
 * Count each row pattern and each column's black pixels. A valid row pattern
 * must occur exactly target times and contain target black pixels; each 'B'
 * column with target black pixels contributes target valid pixels.
 *
 * Time: O(RC). Space: O(RC) for row-pattern strings.
 */

import java.util.HashMap;
import java.util.Map;

// @lc code=start
class Solution {
    public int findBlackPixel(char[][] picture, int target) {
        int rows = picture.length;
        int cols = picture[0].length;
        int[] columnBlack = new int[cols];
        Map<String, Integer> rowCount = new HashMap<>();

        for (char[] row : picture) {
            String pattern = new String(row);
            rowCount.merge(pattern, 1, Integer::sum);
            for (int c = 0; c < cols; c++) {
                if (row[c] == 'B') {
                    columnBlack[c]++;
                }
            }
        }

        int answer = 0;
        for (Map.Entry<String, Integer> entry : rowCount.entrySet()) {
            String pattern = entry.getKey();
            int occurrences = entry.getValue();
            if (occurrences != target || countBlack(pattern) != target) {
                continue;
            }
            for (int c = 0; c < cols; c++) {
                if (pattern.charAt(c) == 'B' && columnBlack[c] == target) {
                    answer += target;
                }
            }
        }
        return answer;
    }

    private int countBlack(String pattern) {
        int count = 0;
        for (int i = 0; i < pattern.length(); i++) {
            if (pattern.charAt(i) == 'B') {
                count++;
            }
        }
        return count;
    }
}
// @lc code=end

File: pt-1/533.lonely-pixel-ii.go
package leetcode

//
// @lc app=leetcode id=533 lang=golang
//
// [533] Lonely Pixel II
//
// Notes
// Count black pixels in every column and count identical row patterns. A row
// pattern contributes only when it appears exactly target times and contains
// exactly target black pixels; for each black column whose total count is also
// target, all target identical rows contribute. Time: O(RC). Space: O(RC) for
// row-pattern strings.
//
// @lc code=start

func FindBlackPixel533(picture [][]byte, target int) int {
	rows, cols := len(picture), len(picture[0])
	columnBlackCount := make([]int, cols)
	rowCount := map[string]int{}

	for _, row := range picture {
		rowPattern := string(row)
		rowCount[rowPattern]++
		for col, value := range row {
			if value == 'B' {
				columnBlackCount[col]++
			}
		}
	}

	answer := 0
	for rowPattern, occurrences := range rowCount {
		if occurrences != target {
			continue
		}
		if countByte533(rowPattern, 'B') != target {
			continue
		}
		for col := 0; col < cols; col++ {
			if rowPattern[col] == 'B' && columnBlackCount[col] == target {
				answer += target
			}
		}
	}
	_ = rows
	return answer
}

func countByte533(s string, target byte) int {
	count := 0
	for i := range s {
		if s[i] == target {
			count++
		}
	}
	return count
}

// @lc code=end
